deepsubtree.go

```
@@ -0,0 +1,421 @@
+package iavl
+
+import (
+	"bytes"
+	"encoding/binary"
+	"errors"
+	"fmt"
+	"strings"
+
+	ics23 "github.com/confio/ics23/go"
+)
+
+const (
+	// lengthByte is the length prefix prepended to each of the sha256 sub-hashes
+	lengthByte byte = 0x20
+)
+
+// Represents a IAVL Deep Subtree that can contain
+// a subset of nodes of an IAVL tree
+type DeepSubTree struct {
+	*MutableTree
+}
+
+// Traverses the nodes in the NodeDB that are part of Deep Subtree
+// and links them together using the populated left and right
+// hashes and sets the root to be the node with the given rootHash
+func (dst *DeepSubTree) BuildTree(rootHash []byte) error {
+	if dst.root == nil {
+		rootNode, rootErr := dst.ndb.GetNode(rootHash)
+		if rootErr != nil {
+			return fmt.Errorf("could not set root of deep subtree: %w", rootErr)
+		}
+		dst.root = rootNode
+	} else if !bytes.Equal(dst.root.hash, rootHash) {
+		return fmt.Errorf(
+			"deep Subtree rootHash: %s does not match expected rootHash: %s",
+			dst.root.hash,
+			rootHash,
+		)
+	}
+	nodes, traverseErr := dst.ndb.nodes()
+	if traverseErr != nil {
+		return fmt.Errorf("could not traverse nodedb: %w", traverseErr)
+	}
+	for _, node := range nodes {
+		pnode, err := dst.ndb.GetNode(node.hash)
+		if err != nil {
+			return err
+		}
+		err = dst.linkNode(pnode)
+		if err != nil {
+			return err
+		}
+	}
+	// Now that nodes are linked correctly, traverse again
+	// and set their keys correctly
+	for _, node := range nodes {
+		pnode, err := dst.ndb.GetNode(node.hash)
+		if err != nil {
+			return err
+		}
+		if pnode.leftNode != nil {
+			pnode.key = pnode.leftNode.getHighestKey()
+		}
+
+		if pnode.rightNode != nil {
+			pnode.key = pnode.rightNode.getLowestKey()
+		}
+	}
+
+	return nil
+}
+
+// Link the given node if it is not linked yet
+// If already linked, return an error in case connection was made incorrectly
+// Note: GetNode returns nil if the hash passed into it is empty which is expected
```

I don't think this is a valid reason to ignore the error because this doesn't handles the case of a corrupt hash and nil node, only the case when node is nil and hash is nil, so `GetNode` is returning `nil, ErrNodeMissingHash`. But we would never hit that case because we are checking for `> 0` at the start. So we know that if we are calling `GetNode` we are calling it with a non empty hash. 

Therefore the errors we should be getting from `GetNode` should be an error getting the node or making the node, both of which seem like errors we should care about.

---

Updated the comment. The error starting with `Value missing for hash` is expected with a deep subtree since it does not contain all nodes but can contain left hashes.  `ErrMissingHash` will never be true since we are checking `> 0` like you said.